cupy/core/include/cupy/carray.cuh

```diff
@@ -180,6 +180,7 @@ public:
 private:
   T* data_;
   ptrdiff_t size_;
+  ptrdiff_t c_contiguous;
```

Comment:
How about passing `c_contiguous` as a template parameter?

Reply:
Great idea!!

Reply:
Oh, I happen to be looking into the internals, so while waiting for thoughts from @emcastillo, I can share my objection: If `c_contiguous` is a template parameter instead of a class member, you'd be re-compiling kernels for C-contiguous inputs, reducing the chances of reusing existing kernels. This is a pure waste, since all the kernels do not have any specialization for such cases.

Reply:
oops...we replied in sync 😛

Reply:
I don't think it will create that much additional kernels. 
I guess I can try to run the test-suite and measure the number of created kernels in both cases. something like an extra 15~20% should be admissible I guess

Reply:
Well while this increase not a big deal, I wanna say most of our test cases only use C-contiguous inputs, so it could be possible to have <10% increase 😂

Reply:
Kernels with several parameters can create a combinatorial explosion, so we can end up with one or two orders of magnitude more kernels than the original numbers. I see your point.

Reply:
I tried the template, the speedup is noticeable
```
1570.192
```
Exactly the same as using #3179

Reply:
Yes, the condition in [L219](https://github.com/cupy/cupy/pull/3191/files#diff-8086e46284b5ae516b8e5a426470dac6R219) can be decided at compile-time, so it is good for compile-time optimization.

Reply:
Do you think that the total increase in the number of compiled kernels can be a burden?

Reply:
Currently, `CArray` takes `ndim` as a template parameter, but I feel it is not a burden. So I guess that `c_contiguous` will not be so too.

Reply:
pushed the changes, PTAL

Reply:
Thanks for the checks, Emilio!